### PR TITLE
feat: support inline LU section editing

### LINKED ISSUE #1980: [LU Editor] support inline LU section editing
## Is your feature request related to a problem? Please describe.

To support LU intent (section) editing in form, visual, we need implement Inline-LU Editor base on recently integrated LU LSP.

### Areas:
1. LU all up view intent editing.
2. Visual page new trigger creating.
3. Form editor inputs components.

### Inline features:
- LU syntax highlight.
- LU content validation.
- SubSection / Nested section editing.

### Related works:
https://github.com/microsoft/BotFramework-Composer/pull/1786 @liweitian 
https://github.com/microsoft/BotFramework-Composer/issues/1641

https://github.com/microsoft/BotFramework-Composer/issues/1848 @alanlong9278 
https://github.com/microsoft/BotFramework-Composer/issues/1598

### PR DESCRIPTION
**Pull request by zhixzhan**
## Description

Add support  for LU inline editing in form, visual, we need implement Inline-LU Editor base on recently integrated LU LSP.

##### Inline features:
- LU syntax highlight.
- LU content validation.
- SubSection / Nested section editing.

## Task Item

close #1980

## Screenshots

Simple section editing:
Untitled

Nested section editing:
Untitled

Inline highlighting & suggestion
Screen Shot 2020-02-25 at 12 25 16 PM

### CODE REVIEW

File: Composer/packages/client/src/router.tsx [hunk 1]
```diff
@@ -48,7 +48,8 @@ const Routes = props => {
                 <Router basepath={BASEPATH} {...parentProps}>
                   <Redirect from="/" to={resolveToBasePath(BASEPATH, 'dialogs/Main')} noThrow />
                   <SettingPage path="setting/*" />
-                  <LUPage path="language-understanding/*" />
+                  <Redirect from="language-understanding" to="language-understanding/all" noThrow />
```
Comment (liweitian) [inline Composer/packages/client/src/router.tsx]: do we need to check if a lu file exist? Currently, if the url is language-understanding/hhh, it still shows the language-understanding page. probably 404 or redirect to lanugage-understanding/all?

File: Composer/packages/client/src/pages/language-understanding/code-editor.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+  useEffect(() => {
+    // reset content with file.content's initial state
+    if (!file || isEmpty(file) || content) return;
+    const value = intent ? intent.Body : file.content;
+    setContent(value);
+  }, [file, sectionId]);
+
+  useEffect(() => {
+    // const currentDiagnostics = inlineMode && intent ? filterTemplateDiagnostics(diagnostics, intent) : diagnostics;
```
Comment (liweitian) [inline Composer/packages/client/src/pages/language-understanding/code-editor.tsx]: delete comment code

File: Composer/packages/client/src/router.tsx (hunk 1 above)
Comment (zhixzhan) [reply Composer/packages/client/src/router.tsx]: Nice catch, sure will update it

**Comment by boydc2014** [comment]
I've been playing with this, looks good, except one issue

In inline lu mode, the auto-completion is only complete for entity definition for "this part", but what we really want is complete for all entities defined in this lu file(and the imported ones). 

@cosmicshuai  can you also help take a look

**Comment by cosmicshuai** [comment]
> I've been playing with this, looks good, except one issue
> 
> In inline lu mode, the auto-completion is only complete for entity definition for "this part", but what we really want is complete for all entities defined in this lu file(and the imported ones).
> 
> @cosmicshuai can you also help take a look

Sure. I will look into it.

File: Composer/packages/client/src/pages/language-understanding/code-editor.tsx [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    const value = intent ? intent.Body : file.content;
+    setContent(value);
+  }, [file, sectionId]);
+
+  useEffect(() => {
+    const currentDiagnostics = inlineMode && intent ? filterTemplateDiagnostics(diagnostics, intent) : diagnostics;
+    const isInvalid = !isValid(currentDiagnostics);
+    const text = isInvalid ? combineMessage(diagnostics) : '';
+    setErrorMsg(text);
+  }, [diagnostics]);
```
Comment (yeze322) [inline Composer/packages/client/src/pages/language-understanding/code-editor.tsx]: About the useEffect hook here.

`diagnostics` comes from local state and is always changed by `setDiagnostics`;
`errorMsg` always changes along with `diagnostic` and there is no asyn transformation.

If so, `errorMsg` could be treated as a computed value out of `diagnostics`. Use `useMemo` hook instead of `useEffect` will have better perf.

File: Composer/packages/tools/language-servers/language-understanding/package.json [hunk 4]
```diff
@@ -4,22 +4,22 @@
...
-      "lint": "eslint --quiet --ext .js,.jsx,.ts,.tsx ./src ./__tests__",
-      "lint:fix": "yarn lint --fix",
-      "lint:typecheck": "tsc --noEmit",
-      "startDemo": "concurrently --kill-others \"yarn start:server\" \"yarn start:client\"",
-      "start:client": "cd demo && webpack-dev-server",
-      "start:server": "cd demo && ts-node ./src/server.ts"
+        "build": "tsc --build tsconfig.build.json",
+        "build:demo": "cd demo && tsc --build tsconfig.json",
+        "prepublishOnly": "npm run build",
+        "clean": "rimraf lib demo/dist",
```
Comment (yeze322) [inline Composer/packages/tools/language-servers/language-understanding/package.json]: these are unnecessary changes

File: Composer/packages/client/src/pages/language-understanding/code-editor.tsx (hunk 3 above)
Comment (yeze322) [reply Composer/packages/client/src/pages/language-understanding/code-editor.tsx]: The displayed `'errorMsg'` consists of two parts, one computed from `diagnostics`, another yielded from catched error. They can be managed separately and joined together as the final value.
```
errorMsg={computedFromDiagnostics || state.errorMsg}
```

Comment (zhixzhan) [reply Composer/packages/client/src/pages/language-understanding/code-editor.tsx]: Updated, great review!
